File: ios/Gekidou/Sources/Gekidou/SessionAttributes/SessionAttributesCollector.swift
//
//  SessionAttributesCollector.swift
//  Gekidou
//
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
//

import Foundation
import Network
import NetworkExtension
import SystemConfiguration.CaptiveNetwork

struct NetworkSnapshot {
    let interfaceType: String
    let ipAddress: String
    let vpnActive: Bool
    let ssid: String
}

public class SessionAttributesCollector {
    public static let shared = SessionAttributesCollector()

    private let store = SessionAttributesStore.shared
    private var fqdnCache: [String: String] = [:]

    private let monitorQueue = DispatchQueue(label: "com.mattermost.sessionattributes.network")
    private let pathMonitor = NWPathMonitor()
    private var cachedInterfaceType = ""
    private var cachedSsid = ""
    private let ssidUnavailable: Bool

    private init() {
        ssidUnavailable = Bundle.main.bundlePath.hasSuffix(".appex") &&
            Bundle.main.bundleIdentifier?.contains("NotificationService") == true

        pathMonitor.pathUpdateHandler = { [weak self] path in
            guard let self else {
                return
            }
            if path.usesInterfaceType(.wifi) {
                self.cachedInterfaceType = "wifi"
            } else if path.usesInterfaceType(.cellular) {
                self.cachedInterfaceType = "cellular"
            } else if path.usesInterfaceType(.wiredEthernet) {
                self.cachedInterfaceType = "ethernet"
            } else if path.status == .unsatisfied {
                self.cachedInterfaceType = ""
            } else {
                self.cachedInterfaceType = "other"
            }

            if self.cachedInterfaceType == "wifi" {
                self.refreshSsid()
            } else {
                self.cachedSsid = ""
            }
        }
        pathMonitor.start(queue: monitorQueue)
    }

    func collect(_ name: String, serverUrl: String) -> String {
        if let stableValue = store.getStableValue(name), !stableValue.isEmpty {
            return stableValue
        }

        switch name {
        case SessionAttributesConstants.AttributeKey.clientIpAddress:
            return currentNetworkSnapshot().ipAddress
        case SessionAttributesConstants.AttributeKey.networkInterfaceType:
            return currentNetworkSnapshot().interfaceType
        case SessionAttributesConstants.AttributeKey.vpnActive:
            return currentNetworkSnapshot().vpnActive ? "true" : "false"
        case SessionAttributesConstants.AttributeKey.ssid:
            return currentNetworkSnapshot().ssid
        case SessionAttributesConstants.AttributeKey.mdmEnrolled:
            return isMdmEnrolled() ? "true" : "false"
        case SessionAttributesConstants.AttributeKey.serverFqdn:
            return getServerFqdn(serverUrl)
        default:
            return ""
        }
    }

    private func getServerFqdn(_ serverUrl: String) -> String {
        if let cached = fqdnCache[serverUrl] {
            return cached
        }
        guard let url = URL(string: serverUrl), let host = url.host, !host.isEmpty else {
            fqdnCache[serverUrl] = ""
            return ""
        }
        fqdnCache[serverUrl] = host
        return host
    }

    private func isMdmEnrolled() -> Bool {
        guard let managedConfig = UserDefaults.standard.dictionary(forKey: SessionAttributesConstants.managedConfigKey) else {
            return false
        }
        return isTruthyManagedFlag(managedConfig[SessionAttributesConstants.isDeviceManagedKey]) ||
            isTruthyManagedFlag(managedConfig[SessionAttributesConstants.isSupervisedKey])
    }

    private func isTruthyManagedFlag(_ value: Any?) -> Bool {
        guard let value else {
            return false
        }
        if let bool = value as? Bool {
            return bool
        }
        if let number = value as? NSNumber {
            return number.boolValue
        }
        if let string = value as? String {
            let normalized = string.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
            return normalized == "true" || normalized == "1" || normalized == "yes"
        }
        return false
    }

    func currentNetworkSnapshot() -> NetworkSnapshot {
        let vpnActive = hasVpnInterface()
        let ipAddress = resolveIpAddress()
        return monitorQueue.sync {
            let interfaceType = vpnActive ? "vpn" : cachedInterfaceType
            let ssid = interfaceType == "wifi" ? cachedSsid : ""
            return NetworkSnapshot(
                interfaceType: interfaceType,
                ipAddress: ipAddress,
                vpnActive: vpnActive,
                ssid: ssid
            )
        }
    }

    private func refreshSsid() {
        guard !ssidUnavailable else {
            cachedSsid = ""
            return
        }

        if #available(iOS 14.0, *) {
            NEHotspotNetwork.fetchCurrent { [weak self] network in
                guard let self else {
                    return
                }
                let ssid: String
                if let networkSSID = network?.ssid,
                   networkSSID != "Wi-Fi",
                   networkSSID != "WLAN" {
                    ssid = networkSSID
                } else {
                    ssid = ""
                }
                self.monitorQueue.async {
                    self.cachedSsid = ssid
                }
            }
            return
        }

        cachedSsid = resolveLegacySsid()
    }

    private func resolveLegacySsid() -> String {
        guard let interfaces = CNCopySupportedInterfaces() as? [String] else {
            return ""
        }
        for interface in interfaces {
            if let info = CNCopyCurrentNetworkInfo(interface as CFString) as? [String: AnyObject],
               let networkSSID = info[kCNNetworkInfoKeySSID as String] as? String,
               networkSSID != "Wi-Fi",
               networkSSID != "WLAN" {
                return networkSSID
            }
        }
        return ""
    }

    private func hasVpnInterface() -> Bool {
        var interfaces: UnsafeMutablePointer<ifaddrs>?
        guard getifaddrs(&interfaces) == 0, let first = interfaces else {
            return false
        }
        defer { freeifaddrs(interfaces) }

        var ptr = first
        while true {
            let name = String(cString: ptr.pointee.ifa_name)
            if name.hasPrefix("utun") || name.hasPrefix("ipsec") || name.hasPrefix("ppp") {
                return true
            }
            guard let next = ptr.pointee.ifa_next else {
                break
            }
            ptr = next
        }
        return false
    }

    private func resolveIpAddress() -> String {
        var address = ""
        var interfaces: UnsafeMutablePointer<ifaddrs>?
        guard getifaddrs(&interfaces) == 0, let first = interfaces else {
            return address
        }
        defer { freeifaddrs(interfaces) }

        var ptr = first
        while true {
            let interface = ptr.pointee
            if let ifaAddr = interface.ifa_addr, ifaAddr.pointee.sa_family == UInt8(AF_INET) {
                let name = String(cString: interface.ifa_name)
                if name == "en0" || name == "en1" {
                    var addr = ifaAddr.pointee
                    var buffer = [CChar](repeating: 0, count: Int(INET_ADDRSTRLEN))
                    inet_ntop(AF_INET, &addr, &buffer, socklen_t(INET_ADDRSTRLEN))
                    let ip = String(cString: buffer)
                    if ip != "0.0.0.0" {
                        address = ip
                    }
                }
            }
            guard let next = interface.ifa_next else {
                break
            }
            ptr = next
        }
        return address
    }
}

File: libraries/@mattermost/rnutils/android/src/main/java/com/mattermost/rnutils/session_attributes/SessionAttributesCollector.kt
package com.mattermost.rnutils.session_attributes

import android.app.admin.DevicePolicyManager
import android.content.Context
import android.content.RestrictionsManager
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.net.wifi.WifiManager
import android.os.Build
import java.net.Inet4Address
import java.net.NetworkInterface
import java.util.concurrent.ConcurrentHashMap

data class NetworkSnapshot(
    val interfaceType: String,
    val ipAddress: String,
    val vpnActive: Boolean,
    val ssid: String,
)

class SessionAttributesCollector(
    private val context: Context,
    private val store: SessionAttributesStore,
) {
    private val fqdnCache = ConcurrentHashMap<String, String>()

    fun collect(name: String, serverUrl: String): String {
        store.getStableValue(name)?.let { return it }

        val snapshot = currentNetworkSnapshot()
        return when (name) {
            SessionAttributesConstants.AttributeKey.CLIENT_IP_ADDRESS -> snapshot.ipAddress
            SessionAttributesConstants.AttributeKey.NETWORK_INTERFACE_TYPE -> snapshot.interfaceType
            SessionAttributesConstants.AttributeKey.VPN_ACTIVE -> if (snapshot.vpnActive) "true" else "false"
            SessionAttributesConstants.AttributeKey.SSID -> snapshot.ssid
            SessionAttributesConstants.AttributeKey.MDM_ENROLLED -> if (isMdmEnrolled()) "true" else "false"
            SessionAttributesConstants.AttributeKey.SERVER_FQDN -> getServerFqdn(serverUrl)
            else -> ""
        }
    }

    private fun getServerFqdn(serverUrl: String): String {
        return fqdnCache.computeIfAbsent(serverUrl) { url ->
            try {
                android.net.Uri.parse(url).host ?: ""
            } catch (_: Exception) {
                ""
            }
        }
    }

    private fun isMdmEnrolled(): Boolean {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            val devicePolicyManager = context.getSystemService(Context.DEVICE_POLICY_SERVICE) as? DevicePolicyManager
            if (devicePolicyManager?.isOrganizationOwnedDeviceWithManagedProfile == true) {
                return true
            }
        }

        val restrictionsManager = context.getSystemService(Context.RESTRICTIONS_SERVICE) as? RestrictionsManager
        val restrictions = restrictionsManager?.applicationRestrictions
        if (restrictions != null) {
            if (isTruthyManagedFlag(restrictions.get(SessionAttributesConstants.IS_DEVICE_MANAGED_KEY)) ||
                isTruthyManagedFlag(restrictions.get(SessionAttributesConstants.IS_SUPERVISED_KEY))
            ) {
                return true
            }
        }
        return false
    }

    private fun isTruthyManagedFlag(value: Any?): Boolean {
        return when (value) {
            is Boolean -> value
            is Int -> value != 0
            is String -> {
                val normalized = value.trim().lowercase()
                normalized == "true" || normalized == "1" || normalized == "yes"
            }
            else -> false
        }
    }

    fun currentNetworkSnapshot(): NetworkSnapshot {
        val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val activeNetwork = connectivityManager.activeNetwork
        val capabilities = activeNetwork?.let { connectivityManager.getNetworkCapabilities(it) }

        val vpnActive = capabilities?.hasTransport(NetworkCapabilities.TRANSPORT_VPN) == true
        val interfaceType = when {
            vpnActive -> "vpn"
            capabilities?.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) == true -> "wifi"
            capabilities?.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) == true -> "cellular"
            capabilities?.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET) == true -> "ethernet"
            activeNetwork == null -> ""
            else -> "other"
        }

        val ipAddress = resolveIpAddress()
        val ssid = if (interfaceType == "wifi") resolveSsid() else ""

        return NetworkSnapshot(interfaceType, ipAddress, vpnActive, ssid)
    }

    private fun resolveIpAddress(): String {
        return try {
            val interfaces = NetworkInterface.getNetworkInterfaces()
            while (interfaces.hasMoreElements()) {
                val networkInterface = interfaces.nextElement()
                val addresses = networkInterface.inetAddresses
                while (addresses.hasMoreElements()) {
                    val address = addresses.nextElement()
                    if (!address.isLoopbackAddress && address is Inet4Address) {
                        return address.hostAddress ?: ""
                    }
                }
            }
            ""
        } catch (_: Exception) {
            ""
        }
    }

    @Suppress("DEPRECATION")
    private fun resolveSsid(): String {
        return try {
            val wifiManager = context.applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
            val info = wifiManager.connectionInfo
            val ssid = info?.ssid?.replace("\"", "") ?: ""
            if (ssid == "<unknown ssid>" || ssid == "0x" || ssid == "Wi-Fi" || ssid == "WLAN") {
                ""
            } else {
                ssid
            }
        } catch (_: Exception) {
            ""
        }
    }
}
